jakarta-cdi/src/main/java/io/cucumber/jakarta/cdi/CdiJakartaFactory.java

```diff
@@ -69,4 +76,51 @@ public boolean addClass(final Class<?> clazz) {
         }
         return selected.get();
     }
+
+    void afterBeanDiscovery(final @Observes AfterBeanDiscovery afterBeanDiscovery, BeanManager bm) {
+        Set<Class<?>> unmanagedClasses = new HashSet<>();
+
+        for (Class<?> stepClass : stepClasses) {
+            discoverUnmanagedClasses(afterBeanDiscovery, bm, unmanagedClasses, stepClass);
+        }
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    void discoverUnmanagedClasses(
+            AfterBeanDiscovery afterBeanDiscovery, BeanManager bm, Set<Class<?>> unmanagedClasses,
+            Class<?> clazz
+    ) {
+        if (!unmanagedClasses.contains(clazz) && bm.getBeans(clazz).isEmpty()) {
+            unmanagedClasses.add(clazz);
+
+            final InjectionTarget injectionTarget = addBean(afterBeanDiscovery, bm, clazz);
+
+            Set<InjectionPoint> ips = injectionTarget.getInjectionPoints();
+            for (InjectionPoint ip : ips) {
+                Type type = ip.getType();
+                if (type instanceof Class) {
```

Comment:
What happens if I try to inject an unmanaged parametrized type e.g. `KafkaProducer<String, Pojo>`?

Reply:
I've added a warning if/when this happens and the parameterized bean isn't managed by having a beans.xml. Though that makes me think it should be possible to do so. Not necessary to fix right now though. Some one else can do that.

Reply:
This is supported by cdi so no additional warning is needed

Reply:
Currently the implementation uses:

```
AnnotatedType clazzAnnotatedType = beanManager.createAnnotatedType(clazz);
```

This only accepts class types and does not support parametrized types. Could you propose a change that makes it work?

Reply:
Good catch with parameterized types, I did not think about it.
Actually a bean can satisfies multiple types and there's a whole set of rules about which bean can satisfies the injection point.

I've modified the code to support parameterized types and adjusted the tests, but I think we need more tests to make sure it works correctly, what I have in mind is having multiple injection points (in multiple beans) but with different parameterized types and also without any types at all.

With the current code I think we will inject a bean for each and I want to make sure we will not get a AmbiguousDependency exception.

I'll try to add those tests tonight or tomorrow.

Reply:
addBean() is the entry point, you set beanClass(rawType).types(paramType, Object.class).scope(Default.Literal.INSTANCE, Any.Literal.INSTANCE).scope(Dependent.class).id(whatyoulike).createWith(c -> ...).
Types is what is used, bean class is mainly for proxies when needed - generally not here.

Reply:
Also it is quite safe to add unmanaged beans for step classes but as soon as we resolve new types to satisfy injection points we are bound to get some unthought situation. 

I think it would be safer to log a message and let the user know we recommend adding a `beans.xml` file (or enable bean discovery, there might be a `beans.xml` already).

Reply:
the standard trick is to add a qualifier on the injection point (@CucumberInternal for ex) you handle and add it as well when registering the bean, avoids ambiguous situations. To be concrete any warning in a CDI env = a deployment error so logging it is pointless normally.

Reply:
Good points. I'd like to prevent is a situation where we create a passing test but a failing application by creating beans for all unsatisfied dependencies.

We're currently recursively going through the step definitions and their dependencies. I'm thinking that perhaps instead of recursively adding all dependencies, we only add the step definitions and check if the dependencies can be injected. If not, then we can throw an informative error.

Reply:
I think we should log a warning as soon as we create a bean definition to let the user know it would be better to add a `beans.xml` (and if necessary create Producers) to take full control.

One scenario I was thinking was a user (let's call him John) including a shared test-jar with a simple Step class that `@Inject`s a POJO. Let's say they were using guice to do the injection so they did not include a `beans.xml` file in the test-jar. 

Now John cannot simply add a `beans.xml` because the step class comes from a test-jar dependency. He also needs to create a `@Producer`. So even though I'm sure there are some situations that we won't be able to configure correctly, I believe that for all the Johns in the world that are not so familiar with CDI it can still be useful to them if we try to create beans for all the unmanaged injection points.

For the more eccentric cases, they will get a warning from us and an error from the CDI container.

Reply:
I wouldnt do something i consider as a warning here. CDI gives you all data to know it will be an error or not so there is no warn case I think. But logging an info or even debug message that a custom bean was registered sounds ok to me.